eth/handler.go

```diff
@@ -361,474 +296,115 @@ func (pm *ProtocolManager) handle(p *peer) error {
-		}
-		if len(transactions) > 0 || len(uncles) > 0 || !filter {
-			err := pm.downloader.DeliverBodies(p.id, transactions, uncles)
-			if err != nil {
-				log.Debug("Failed to deliver bodies", "err", err)
-			}
-		}
+	if err := h.downloader.SnapSyncer.Register(peer); err != nil {
+		return err
+	}
+	// Handle incoming messages until the connection is torn down
+	return handler(peer)
+}
 
-	case p.version >= eth63 && msg.Code == GetNodeDataMsg:
-		// Decode the retrieval message
-		msgStream := rlp.NewStream(msg.Payload, uint64(msg.Size))
-		if _, err := msgStream.List(); err != nil {
-			return err
-		}
-		// Gather state data until the fetch or network limits is reached
-		var (
-			hash  common.Hash
-			bytes int
-			data  [][]byte
-		)
-		for bytes < softResponseLimit && len(data) < downloader.MaxStateFetch {
-			// Retrieve the hash of the next state entry
-			if err := msgStream.Decode(&hash); err == rlp.EOL {
-				break
-			} else if err != nil {
-				return errResp(ErrDecode, "msg %v: %v", msg, err)
-			}
-			// Retrieve the requested state entry, stopping if enough was found
-			// todo now the code and trienode is mixed in the protocol level,
-			// separate these two types.
-			if !pm.downloader.SyncBloomContains(hash[:]) {
-				// Only lookup the trie node if there's chance that we actually have it
-				continue
-			}
-			entry, err := pm.blockchain.TrieNode(hash)
-			if len(entry) == 0 || err != nil {
-				// Read the contract code with prefix only to save unnecessary lookups.
-				entry, err = pm.blockchain.ContractCodeWithPrefix(hash)
-			}
-			if err == nil && len(entry) > 0 {
-				data = append(data, entry)
-				bytes += len(entry)
-			}
-		}
-		return p.SendNodeData(data)
+func (h *handler) removePeer(id string) {
+	// Remove the eth peer if it exists
+	eth := h.peers.ethPeer(id)
+	if eth != nil {
+		log.Debug("Removing Ethereum peer", "peer", id)
+		h.downloader.UnregisterPeer(id)
+		h.txFetcher.Drop(id)
 
-	case p.version >= eth63 && msg.Code == NodeDataMsg:
-		// A batch of node state data arrived to one of our previous requests
-		var data [][]byte
-		if err := msg.Decode(&data); err != nil {
-			return errResp(ErrDecode, "msg %v: %v", msg, err)
+		if err := h.peers.unregisterEthPeer(id); err != nil {
+			log.Error("Peer removal failed", "peer", id, "err", err)
 		}
-		// Deliver all to the downloader
-		if err := pm.downloader.DeliverNodeData(p.id, data); err != nil {
-			log.Debug("Failed to deliver node state data", "err", err)
+	}
+	// Remove the snap peer if it exists
+	snap := h.peers.snapPeer(id)
+	if snap != nil {
+		log.Debug("Removing Snapshot peer", "peer", id)
+		h.downloader.SnapSyncer.Unregister(id)
+		if err := h.peers.unregisterSnapPeer(id); err != nil {
+			log.Error("Peer removal failed", "peer", id, "err", err)
 		}
+	}
+	// Hard disconnect at the networking layer
+	if eth != nil {
+		eth.Peer.Disconnect(p2p.DiscUselessPeer)
+	}
+	if snap != nil {
```

Comment:
is there a reason for doing the additional check for `snap != nil` down here too? 

can you disconnect useless peer after unregistering the peer after L346?